server/tests/mod.rs

```diff
@@ -72,6 +72,37 @@ async fn send_message() {
         .await;
 }
 
+#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+#[tracing::instrument(name = "Rate limit test", skip_all)]
+async fn rate_limit() {
+    let mut setup = TestBackend::single_with_rate_limits(RateLimitsConfig {
+        period: Duration::from_secs(100), // technically, never replenish tokens
+        burst_size: 2,                    // allow in total 2 request
+    })
+    .await;
+    setup.add_user(&ALICE).await;
+    setup.add_user(&BOB).await;
+    let conversation_id = setup.connect_users(&ALICE, &BOB).await;
+
+    let alice = setup.users.get_mut(&ALICE).unwrap();
+    let error = alice
+        .user
+        .send_message(
+            conversation_id,
+            MimiContent::simple_markdown_message("Hello bob".into()),
+        )
+        .await
+        .unwrap_err();
+
+    let error: DsRequestError = error.downcast().expect("should be a DsRequestError");
+    match error {
+        DsRequestError::Tonic(status) => {
+            assert_eq!(status.code(), tonic::Code::ResourceExhausted);
+        }
+        _ => panic!("unexpected error type: {:?}", error),
+    }
+}
```

Comment:
I think the test is a bit brittle because it assumes the setup phase won't run into the rate-limiting. Maybe a higher limit would be better – and then you trigger it with sending messages in a loop.

Reply:
It would also be good to test the replenishing. Choosing a short enough period and then just sleeping until it's over should be doable.

Reply:
Good idea. Both done.